---
title: "RNAseq with edgeR"
author: "Jakob Jung"
date: "November 17, 2020"
output:
  pdf_document: default
  html_document:
    df_print: paged
---
Here I perform the downstream analysis of the RNAseq experiment, in which EHEC was challenged with PNA targeting the different genes. RNAseq was performed to determine transcriptome changes upon exposure to PNA. 

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE, warning = TRUE)
```

# Packages
I import the packages needed for all analysis. They can all be installed from Bioconductor or CRAN if not statet otherwise:
```{r, message=FALSE}
library(edgeR)
library(circlize)
library(dplyr)
library(ggplot2)
library('RUVSeq')
library(RColorBrewer)
library(oligo)
library(EDASeq)
library(gplots)
library(ggrepel)
library(svglite)
library(ComplexHeatmap)
library(VennDiagram)
library(eulerr)
library(tidyverse)
library(grid)

```


# Data Acquisition  

I generated a tab file containing gene counts after upstream processing. Upstream processing included folowing steps (starting with fastq-files): 
 
 - BBtools for filtering, trimming and mapping 
 - featureCounts for generating a count matrix


Firstly, I import the gene-wise counts:
```{r}
GenewiseCounts <- read.delim(
  "../data/rna_align/counttable.txt",sep = "\t",
  row.names = 1, header = T, comment.char = "#")

dim(GenewiseCounts)
head(GenewiseCounts[,1:6])
```
\hfill\break

I have to change column names, since they include the whole path:
```{r}
gwc <- GenewiseCounts[,5:length(GenewiseCounts[1,])]
pnapat <- "\\.\\.\\.data\\.rna_align\\..*_(\\d)_(.*)\\.bam"
colnames (gwc) <- gsub(pnapat,"\\2_\\1", colnames(gwc))
colnames (gwc)
newnames <- read.delim("../data/samples/samples.tsv", header = F) 
newnames <- unlist(newnames)
names(newnames) <- NULL
newnames <- gsub("-", "_", newnames)
colnames(gwc) <- c("Length",newnames)
```

I also create a facor variable for groups of the sample data manually (from assigning sample codes to condition):
```{r}
test <- gsub("_\\d", "" , newnames)
test <- as.factor(test)
test
```

Now that I have the read count dataframe with sample names, I import them into the edgeR environment:
```{r}
y <- DGEList(gwc[,-1], group = test, genes = gwc[,1,drop=FALSE])
options(digits = 3)
head(y$samples)
```


# Filtering
Now I want to filter out Genes which have very low counts across all libraries. 
I do this by creating a cutoff $$\frac {10} {L} $$
where L is the minimum library size in millions. We delete genes that are below the cutoff in at least 2 libraries:
```{r}
L <- min(y$samples$lib.size) / 1000000
cutoff <- 10/L
keep <- rowSums(cpm(y) > cutoff) >= 2 
table(keep)
```
I retain only the unfiltered genes,and delete 519 genes below the threshold:
```{r}
y <- y[keep, , keep.lib.sizes=FALSE]
```



# Design matrix
I create a design matrix for the samples:
```{r}
batches <- factor(gsub(".*(\\d).*","\\1",rownames(y$samples)))
design <- model.matrix(~0+test+batches)
colnames(design) <- c(levels(test), "batch")
rownames(design) <- colnames(y$counts)
design[1:5,]
```


# Normalization

I check how the standard TMM normalization of edgeR performs. I start with calculating normalization factors:
```{r, error=FALSE}
y <- calcNormFactors(y)
y <- estimateDisp(y, design, robust = T)
```

And now I create PCA and RLE plots:
```{r}
library(RColorBrewer)
nb.cols <- 22
mycolors <- colorRampPalette(brewer.pal(8, "Set1"))(nb.cols)


plotPCA(cpm(y), col="black", bg=mycolors[test],  labels=F, pch=21, cex=3)
legend("bottomright", legend = levels(test), pch=21, 
       cex=1.2, col="black",  pt.bg = mycolors[test], pt.cex = 2)

plotPCA(cpm(y), col="black", bg=mycolors[test],  labels=F, pch=21, cex=3)
legend("bottomright", legend = levels(test), pch=21, 
       cex=1.2, col="black",  pt.bg = mycolors[test], pt.cex = 2)

plotPCA(cpm(y), col=mycolors[test], bg=mycolors[test])
plotPCA(cpm(y)[,1:30], col=mycolors[test[1:30]], bg=mycolors[test[1:30]])
plotPCA(cpm(y)[,31:37], col=mycolors[test[31:37]], bg=mycolors[test[31:37]])


plotRLE(cpm(y), outline=FALSE, ylim=c(-1, 1), col=mycolors[test],
        main="RLE")
```

You can see that the TMM was succesful (TMM centers the RLE around 0). 


I save the PCA as pdfs:
```{r}
pdf("../analysis/PCA_TMM_allsamples.pdf")
plotPCA(cpm(y), col=mycolors[test], bg=mycolors[test])
dev.off()
pdf("../analysis/PCA_TMM_mullerhinton.pdf")
plotPCA(cpm(y)[,1:30], col=mycolors[test[1:30]], bg=mycolors[test[1:30]])
dev.off()
pdf("../analysis/PCA_TMM_m9.pdf")
plotPCA(cpm(y)[,31:37], col=mycolors[test[31:37]], bg=mycolors[test[31:37]])
dev.off()
```


# Differential Expression analysis using TMM
Next, I perform differential expression analysis using TMM-normalized dataset:
```{r}
# make a contrast:
con <- makeContrasts(acpP_vs_ctrl = acpP - control,
                     acpP_scr_vs_ctrl = acpP_scr - control,
                     adk_vs_ctrl = adk - control, 
                     csrA_vs_ctrl = csrA - control, 
                     dnaB_vs_ctrl = dnaB - control, 
                     ftsZ_vs_ctrl = ftsZ - control,
                     nusG_vs_ctrl = nusG - control,
                     pyrH_vs_ctrl = pyrH - control,
                     rnp_vs_ctrl = rnp - control,
                     rplS_vs_ctrl = rplS - control,
                     rpoD_vs_ctrl = rpoD - control,
                     rpsH_vs_ctrl = rpsH - control,
                     tnaB_vs_ctrl = tnaB - control,
                     yidC_vs_ctrl = yidC - control,
                     acpP_col_vs_ctrl_col = acpP_col - control_col,
                     acpP_scr_col_vs_ctrl_col = acpP_scr_col - control_col,
                     clbA1_col_vs_ctrl_col = clbA1_col - control_col,
                     clbA2_col_vs_ctrl_col = clbA2_col - control_col,
                     clbP_col_vs_ctrl_col = clbP_col - control_col,
                     clbQ_col_vs_ctrl_col = clbQ_col - control_col,
                     levels = design)

fit <- glmQLFit(y, design, robust = TRUE)

acpP_vs_ctrl <- glmQLFTest(fit, contrast = con[,2])

all_res <- list(acpP_vs_ctrl = glmQLFTest(fit, contrast = con[,1]),
                 acpPscr_vs_ctrl = glmQLFTest(fit, contrast = con[,2]),
                 adk_vs_ctrl = glmQLFTest(fit, contrast = con[,3]), 
                 csrA_vs_ctrl = glmQLFTest(fit, contrast = con[,4]), 
                 dnaB_vs_ctrl = glmQLFTest(fit, contrast = con[,5]), 
                 ftsZ_vs_ctrl = glmQLFTest(fit, contrast = con[,6]),
                 nusG_vs_ctrl = glmQLFTest(fit, contrast = con[,7]),
                 pyrH_vs_ctrl = glmQLFTest(fit, contrast = con[,8]),
                 rnp_vs_ctrl = glmQLFTest(fit, contrast = con[,9]),
                 rplS_vs_ctrl = glmQLFTest(fit, contrast = con[,10]),
                 rpoD_vs_ctrl = glmQLFTest(fit, contrast = con[,11]),
                 rpsH_vs_ctrl = glmQLFTest(fit, contrast = con[,12]),
                 tnaB_vs_ctrl = glmQLFTest(fit, contrast = con[,13]),
                 yidC_vs_ctrl = glmQLFTest(fit, contrast = con[,14]),
                 acpP_col_vs_ctrl_col = glmQLFTest(fit, contrast = con[,15]),
                 acpPscr_col_vs_ctrl_col = glmQLFTest(fit, contrast = con[,16]),
                 clbA1_col_vs_ctrl_col = glmQLFTest(fit, contrast = con[,17]),
                 clbA2_col_vs_ctrl_col = glmQLFTest(fit, contrast = con[,18]),
                 clbP_col_vs_ctrl_col = glmQLFTest(fit, contrast = con[,19]),
                 clbQ_col_vs_ctrl_col = glmQLFTest(fit, contrast = con[,20]))
```




We now create MD, BCV and QLDisp plots to access qualiy of data:
```{r}
plotMD(y, main = "MD-plot")
abline(h=0, col="red", lty=2, lwd=2)
plotBCV(y)
plotQLDisp(fit)
```
The quality looks decent.


Now I create a function which makes nice volcano-plots and run it on all the results (all PNA-samples are compared to water control for DE):
```{r}
do_volcano <- function(restab, targetgene, pointsize = 2, x_limit = F,y_limit=F, show_sig = F, alpha=0.05, 
                       minlogfc=1, title = "Volcano", off_target_list = NULL) {
  rownames(restab) <- gsub("^([^A-Z].+)" , "italic('\\1')" , rownames(restab))
  g = ggplot(restab) +
  geom_point(
    data = restab,
    aes(x = logFC, y = -log10(FDR)),
    color = "darkgrey",
    cex = pointsize
  ) + theme_bw()+ # change theme to standard black&wite.
  geom_hline(yintercept = -log10(alpha),
             color = "black", linetype = 3) +
  geom_vline(xintercept = c(-minlogfc,minlogfc),
             color = "black", linetype = 3) +
  theme(axis.title.x = element_text(size=20),
        axis.title.y = element_text(size=20),
        axis.text = element_text(size=15, colour = "black"),
        panel.background = element_rect(colour = "black"),
        axis.line = element_line(colour = "black"),
        panel.grid.minor.x = element_blank(),
        panel.grid.minor.y = element_blank(),
        panel.grid.major.x =  element_blank(),#element_line(colour="lightgrey", size=0.3),
        panel.grid.major.y = element_blank(),#element_line(colour="lightgrey", size=0.3),
        plot.title = element_text(hjust = 0.5, size = 23))+
  ggtitle(title)+
  xlab(expression("Log"[2]*" fold change")) +
  ylab("- Log10 P-value (FDR)")+
  scale_x_continuous(expand = c(0,0),breaks = seq(-6,6,2), limits = c(-x_limit,x_limit)) +
  scale_y_continuous(expand = c(0, 0),breaks = seq(0,16,2), limits = c(0,y_limit)) 

  
  if (is.null(off_target_list)) {
    g <- g + 
      geom_point(
        data = restab[restab$FDR<alpha & restab$logFC < -minlogfc,],
        aes(x = logFC, y = -log10(FDR)),
        color = "blue", 
        cex = pointsize) +
      geom_point(
        data = restab[restab$FDR<alpha & restab$logFC > minlogfc,],
        aes(x = logFC, y = -log10(FDR)),
        color = "red", 
        cex = pointsize) 
  } else{
    # show all mismatches with significant regulation:
    show <- restab[unlist(off_target_list),][which(restab[unlist(off_target_list),]$FDR < alpha),]
    # different colors for different mismatches:
    g <- g +
      geom_point(
        data = restab[unlist(off_target_list[3]),],
        aes(x = logFC, y = -log10(FDR)),
        color = "darkblue", 
        cex = pointsize) +
      geom_point(
        data = restab[unlist(off_target_list[2]),],
        aes(x = logFC, y = -log10(FDR)),
        color = "blue", 
        cex = pointsize) +
    #zero mismatches:
      geom_point(
        data = restab[unlist(off_target_list[1]),],
        aes(x = logFC, y = -log10(FDR)),
        color = "cyan", 
        cex = pointsize) #+
 #   geom_label_repel(
  #    data = show ,
  #    aes(x = logFC, y = -log10(FDR), 
  #        label = rownames(show)),
  #    hjust = 0.1,
  #    vjust = 2,
   #   size = 4, segment.alpha = 0.5,min.segment.length=0, segment.color = "black") 
  }
  
  # show the sigficantest genes:
  if(show_sig){
    range01 <- function(x){(x-min(x))/(max(x)-min(x))}
    top_up <- restab[ which(restab$FDR < alpha & restab$logFC > minlogfc),]
    top_down <- restab[ which(restab$FDR < alpha & restab$logFC < -(minlogfc)),]
    
    if (length(rownames(top_up)) > 0 && (length(rownames(top_up)) > 3)){
    logFC.scaled <- range01(top_up$logFC)
    FDR.scaled <- range01(-log(top_up$FDR))
    summ <- (logFC.scaled + FDR.scaled)
    top_up <- top_up[order(-summ),][1:3,]
    }

    if (length(rownames(top_down))>0 && (length(rownames(top_down))> 3)){
      logFC.scaled <- range01(-top_down$logFC)
      FDR.scaled <- range01(-log(top_down$FDR))
      summ <- (logFC.scaled + FDR.scaled)
      top_down <- top_down[order(-summ),][1:3,]
    }

    top_peaks <- rbind(top_up, top_down)
    top_peaks <- na.omit(top_peaks)

    g <- g + geom_label_repel(
    data = top_peaks ,
    aes(x = logFC, y = -log10(FDR), 
        label = rownames(top_peaks)),
    hjust = 0.1,
    #vjust = 2, 
    size = 5, segment.alpha = 0.5, segment.color = "black", min.segment.length=unit(0, "cm"), parse = T) 
  }
  g <- g + geom_point(
        data = restab[targetgene,],
        aes(x = logFC, y = -log10(FDR)),
        color = "red", 
        cex = pointsize)
  g
}
```



```{r}
# I get the links between locus tags and gene names:
# pnames <- read.delim("../data/link_lt_gn.tab", header = F)
# rownames(pnames) <- pnames$V2
```


# Off-target analysis:
```{r}
startot <- read.delim("../data/off_targets_search/pna_2mm_startregions.tab")
```

Now I adjust p-values (FDR), create volcano plots, histograms for the results (and save volcano plots as pdfs): 
```{r}
for (resname in names(all_res)){
  # adjust p-values FDR
  all_res[[resname]]$table$FDR <- p.adjust(all_res[[resname]]$table$PValue, method = "fdr")
  restab <- all_res[[resname]]$table
  
  restab$genes <- rownames(restab)
  
  hist(restab$PValue, breaks=100, main=resname)
  
  # check for off-targets (with 0 mismatches):
  targetgene <- gsub("_.*", "" , resname)
  offt_zero_mm <- startot %>% filter(grepl(targetgene, probe_id) & num_mismatch==0) %>% 
    select(trans_id) %>% unlist
  offt_one_mm <- startot %>% filter(grepl(targetgene, probe_id) & num_mismatch==1) %>% 
    select(trans_id) %>% unlist
  offt_two_mm <- startot %>% filter(grepl(targetgene, probe_id) & num_mismatch==2) %>% 
    select(trans_id) %>% unlist
  off_targets <- list(zero = offt_zero_mm, one = offt_one_mm, two = offt_two_mm)
  
  tgene_lt <- gsub(".*(ECP_.+)", "\\1", startot[grepl(targetgene, startot$probe_id),4][1])
  
  # make volcanos:
  pdf(paste("../analysis/volcanoplots/",resname, ".pdf"))
  print(do_volcano(restab, tgene_lt, title=resname, 
                   x_limit = 9,
                   y_limit = 30,
                   alpha=0.001, pointsize = 3, 
                   off_target_list = off_targets))
  dev.off()
  
  #save result_table:
  #dataname <- paste("../analysis/diff_exp_rawdata/", resname, ".csv", sep = "")
  #write.csv(restab[order(restab$FDR),], dataname)
}
```

The volcano-plots that are created look fine, and also the p-value distributions for the samples are uniform with an increment in the lowest p-values showing DE genes.



# Heatmap
I create a heatmap showing the 20 highest ranking DE genesper conditon.
```{r}
all_tables_sorted <- lapply(all_res, function(x) x$table[order(x$table$PValue),]) 


topDEgenes <- sapply(all_tables_sorted, function(x) rownames(x[x$FDR<0.001 & 
                                           abs(x$logFC)>1,])[1:10])
topDEgenes <- unique(topDEgenes[!is.na(topDEgenes)])


logCPM <- cpm(y, prior.count = 2, log = TRUE) 
logCPM <- logCPM[rownames(logCPM)%in%topDEgenes,]
logCPM <- t(scale(t(logCPM))) #centered around 0

# prefname <- ifelse(rownames(logCPM) %in% pnames$V2 ,pnames[rownames(logCPM),]$V1, "" )
# prefname <- ifelse(isUnique(prefname), prefname, "")
# rownames(logCPM) <- ifelse(prefname != "", prefname, rownames(logCPM))


head(logCPM)
```
Interestingly, the patterns of the samples with scrambled PNA and the test sample (PNA11 and 10 resp.) show similar patterns of the DE genes and cluster together closely.

```{r}
# get log2change, between those and noPNA:
logchange <- data.frame(sapply(all_res, function(x) x$table$logFC), 
                        row.names = rownames(all_res$acpP_vs_ctrl$table))

pvals <- data.frame(sapply(all_res, function(x) x$table$FDR<0.001 & abs(x$table$logFC)>1), 
                        row.names = rownames(all_res$acpP_vs_ctrl$table))

#select only significant ones:
logchange <- logchange[rownames(logchange)%in%topDEgenes,]
pvals <- pvals[rownames(pvals)%in%topDEgenes,]

pvals <-sapply(pvals, function(x) ifelse(x, x <- "*",x<-"") )

logchange_T <- t(logchange)
pvals_T <- t(pvals)
```



Now I save the heatmap as pdf:
```{r}

c1 =  circlize::colorRamp2(c(-2, 0, 2), c("blue", "white", "red"))
c2 = circlize::colorRamp2(c(-2, 0, 2), c("darkblue", "white", "green"))

ht4 <- Heatmap(logchange_T, name = "Log2 FC",
               col = c2,
               cluster_rows = F, cluster_columns = F, show_heatmap_legend = F,
               cell_fun = function(j, i, x, y, width, height, fill) {
                 grid.text(sprintf("%.1s", pvals_T[i, j]), x, y)
               }, 
               border = TRUE, #height = unit(3, "cm"),
               row_names_max_width = max_text_width(c(0,0),gp = gpar(fontsize = 0)),
               column_names_rot = 45)
               #rect_gp = gpar(col = "black", lwd = 1))


lgd2 = Legend(col_fun = c2, title = expression("Log"[2]*" FC"), labels_gp = gpar(fontsize = 10),
             title_gp = gpar(fontsize = 15),grid_width =  unit(0.8, "cm"),
             at = c(-2, 0, 2), legend_width = unit(4, "cm"),
             labels = c("-2", "  0", "  2"), legend_height = unit(3, "cm"),
             title_position = "leftcenter-rot")
ht4
```







# KEGG pathway analysis

I perform the KEGG-analysis using the FRY gene set analysis tool from limma:
```{r}
library(KEGGREST)
# get link and list to get kegg info:
link_kegg <- keggLink("pathway", "sey")
list_kegg <- keggList("pathway", "sey")

kegg_pw_ids <- names(list_kegg)

#rename genes, remove ones which arent in our data:
names(link_kegg) <- gsub("sey:(.*)", "\\1", names(link_kegg)) #rename genes as locus tags
link_kegg <- link_kegg[names(link_kegg) %in% rownames(res_KFF$PNAKFF$table)] #remove genes not in data


idx_kegg <- sapply(kegg_pw_ids, function(x){
  x <- unique(names(link_kegg[link_kegg == x])) # choose all genes, except duplucates
})
# add phopq pw to kegg
ppq_raw <- read.delim("../data/PHOPQ.tsv", header = T)

for (c in colnames(ppq_raw)) {
  gs <- ppq_raw[[c]][ppq_raw[[c]]!=""]
  gs_lt <- pnames[pnames$V1 %in% gs,]$V2
  idx_kegg[[c]] <- gs_lt[gs_lt %in% rownames(y$counts)]
}

#ppq <- as.character(ppq_raw$V1)
#phopq <- pnames[pnames$V1 %in% ppq,]$V2

#idx_kegg$PhoPQ <- phopq[phopq %in% rownames(y$counts)] # add PhoPQ genes


#do fry: 
kegg_fry_PNAKFF <- fry(y, idx_kegg, design, contrast=con[,1])
kegg_fry_KFF <- fry(y, idx_kegg, design, contrast=con[,3])
kegg_fry_PNAKFFscr <- fry(y, idx_kegg, design, contrast=con[,2])

kegg_fry_PNARXR <- fry(y, idx_kegg, design, contrast=con[,4])
kegg_fry_RXR <- fry(y, idx_kegg, design, contrast=con[,6])
kegg_fry_PNARXRscr <- fry(y, idx_kegg, design, contrast=con[,5])

kegg_fry_PNATAT <- fry(y, idx_kegg, design, contrast=con[,7])
kegg_fry_TAT <- fry(y, idx_kegg, design, contrast=con[,9])
kegg_fry_PNATATscr <- fry(y, idx_kegg, design, contrast=con[,8])

list_kegg_fry <- list(kegg_fry_PNAKFF=kegg_fry_PNAKFF,kegg_fry_PNAKFFscr=kegg_fry_PNAKFFscr,kegg_fry_KFF=kegg_fry_KFF,
                 kegg_fry_PNARXR=kegg_fry_PNARXR,kegg_fry_PNARXRscr=kegg_fry_PNARXRscr,kegg_fry_RXR=kegg_fry_RXR,
                 kegg_fry_PNATAT=kegg_fry_PNATAT,kegg_fry_PNATATscr=kegg_fry_PNATATscr,kegg_fry_TAT=kegg_fry_TAT)

```

add KEGG terms:
```{r}
for (fryres in names(list_kegg_fry)) {
  list_kegg_fry[[fryres]][["TERM"]] <- ifelse(grepl("path",rownames(list_kegg_fry[[fryres]])),
                                              list_kegg[rownames(list_kegg_fry[[fryres]])],
                                              rownames(list_kegg_fry[[fryres]]))
  list_kegg_fry[[fryres]][["TERM"]] <- gsub("(.*) - Salmonella enterica subsp. enterica serovar Typhimurium SL1344",
                                            "\\1", list_kegg_fry[[fryres]][["TERM"]])
  write.csv(list_kegg_fry[[fryres]], paste("../analysis/analysis_complete/supp_tables/", fryres, ".csv"))
  #list_kegg_fry[[fryres]]["PhoPQ",][["TERM"]] <- "PhoPQ"
}


kegg_frysig <- lapply(list_kegg_fry, function(x) x[x[["FDR"]]<0.05 & x[["NGenes"]]>10,])
kegg_siggos <- c()


for (i in names(kegg_frysig)) {
  print(i)
  print(dim(kegg_frysig[[i]]))
  print(kegg_frysig[[i]][,c(1,2,4,7)])
  kegg_siggos <- c(kegg_siggos, rownames(kegg_frysig[[i]][1:10,]))  # can be modified
}

kegg_siggos <- unique(kegg_siggos[!grepl("NA", kegg_siggos)])


```

Create a heatmap-df  for KEGG:
```{r}
idx_kegg_char <- lapply(idx_kegg, as.character)
# I create a dataframe with mean logFC values for each significant GO-term:
hm_kegg_fry_logfc <- t(as.data.frame(lapply(idx_kegg_char[kegg_siggos], function(x){
  PNAKFF <- median(res_KFF$PNAKFF$table[x,]$logFC)
  PNAKFFscr <- median(res_KFF$PNAKFFscr$table[x,]$logFC)
  KFF <- median(res_KFF$KFF$table[x,]$logFC)
  
  PNARXR <- median(res_RXR$PNARXR$table[x,]$logFC)
  PNARXRscr <- median(res_RXR$PNARXRscr$table[x,]$logFC)
  RXR <- median(res_RXR$RXR$table[x,]$logFC)
  
  PNATAT <- median(res_TAT$PNATAT$table[x,]$logFC)
  PNATATscr <- median(res_TAT$PNATATscr$table[x,]$logFC)
  TAT <- median(res_TAT$TAT$table[x,]$logFC)
  c(PNAKFF, PNAKFFscr, KFF, PNARXR, PNARXRscr, RXR, PNATAT, PNATATscr, TAT)
})))

colnames(hm_kegg_fry_logfc) <- c("KFF-acpP ", "KFF-acpP-scrambled ", "KFF-only ","RXR-acpP ", "RXR-acpP-scrambled ", "RXR-only ",
                            "TAT-acpP ", "TAT-acpP-scrambled ", "TAT-only ")
hm_kegg_fry_logfc <- as.data.frame(hm_kegg_fry_logfc)
rownames(hm_kegg_fry_logfc) <- gsub("\\.", "\\:", rownames(hm_kegg_fry_logfc))
```

make heatmap:
```{r}
hm_kegg_fry_logfc <- hm_kegg_fry_logfc[order(hm_kegg_fry_logfc[,1], decreasing = T),]

pvals <- data.frame(sapply(names(list_kegg_fry), function(x) list_kegg_fry[[x]][rownames(hm_kegg_fry_logfc),"FDR"]), 
                    row.names = rownames(hm_kegg_fry_logfc))

#select only significant ones:
pvals <-sapply(pvals, function(x) ifelse(x<0.05, x <- "*",x<-"") )


keggpws <- list_kegg_fry$kegg_fry_PNAKFF[rownames(hm_kegg_fry_logfc),] [["TERM"]]
rownames(hm_kegg_fry_logfc) <- ifelse(!is.na(keggpws),keggpws, rownames(hm_kegg_fry_logfc) )


ord <- c(rep("KFF", 3), rep("RXR", 3), rep("TAT", 3))
lev <- c("KFF", "RXR", "TAT")
```

plot hm (save as pdf):
```{r}
library(circlize)
col_fun = colorRamp2(c(-1, 0, 1), c("darkblue", "beige", "red"))

nnames <- c("KFF-_acpP_", "KFF-_acpP_-scrambled", "KFF", "RXR-_acpP_", "RXR-_acpP_-scrambled", 
          "RXR", "TAT-_acpP_", "TAT-_acpP_-scrambled", "TAT")

w <- length(hm_kegg_fry_logfc$`KFF-acpP `) # width of plot (nr pws)
h <- 9   # height of plot 

ht_vert <- Heatmap(hm_kegg_fry_logfc, cluster_rows = F, cluster_columns = F,
               name = "GO-analysis", col = col_fun,
               show_heatmap_legend = F, 
               row_title_side = "right", row_title_rot = 0,
               border = TRUE, 
               cell_fun = function(j, i, x, y, width, height, fill) {
                 grid.text(sprintf("%.1s", pvals[i, j]), x, y)
               }, 
               column_names_gp = gpar(fontsize = 10),
               row_names_gp = gpar(fontsize = 10),
               column_split = factor(ord, levels = lev), 
               row_title = NULL,
               row_gap = unit(0.1, "cm"),
               width = unit(5, "cm"), height = unit(20, "cm"),
               column_names_rot = 45)

ht_hor <- Heatmap(t(hm_kegg_fry_logfc), cluster_rows = F, cluster_columns = F,
        name = "GO-analysis", col = col_fun,
        show_heatmap_legend = F, 
        row_title_side = "left", row_title_rot = 0,
        column_names_side = "top",
        border = TRUE, 
        cell_fun = function(j, i, x, y, width, height, fill) {
          grid.text(sprintf("%.1s", t(pvals)[i, j]), x, y)
        }, row_labels = gt_render(nnames),
        row_names_gp = gpar(fontsize = 10),
        #column_names_gp = gpar(fontsize = 10),
        row_split = factor(ord, levels = lev), 
        column_title = NULL,
        column_gap = unit(0.1, "cm"),
        width = unit(w*0.8, "cm"), height = unit(h*0.8, "cm"),
        column_names_rot = 40,
        column_names_gp = gpar(col = c("black", rep("darkgreen",2), rep("black", 2), "darkgreen", rep("black", 9),
                                    "darkgreen", "black", "darkgreen", "black", "darkgreen",rep("black", 3)), 
                               fontsize = 10, fontface =  c("plain", rep("bold",2), rep("plain", 2), "bold", rep("plain", 9),
                                    "bold", "plain", "bold", "plain", "bold",rep("plain", 3))
                              )
        )
        #rect_gp = gpar(col = "black", lwd = 0.1))

ht_hor

lgd = Legend(col_fun = col_fun, title = expression("Median logFC"), direction = "horizontal",
             title_gp = gpar(fontsize = 12),
             at = c(-1, 0, 1), legend_width = unit(4, "cm"))

lgd1 = Legend(col_fun = c1, title = expression("Log CPM"), labels_gp = gpar(fontsize = 10),
             title_gp = gpar(fontsize = 15),
             at = c(-2, 0, 2), legend_width = unit(4, "cm"), grid_width =  unit(0.8, "cm"),
             labels = c("-2", "  0", "  2"), legend_height = unit(3, "cm"),
             title_position = "leftcenter-rot")

pdf("../analysis/gene_set_analysis/hm_KEGG_collapsed_new.pdf", width = unit(w*0.45, "cm"))
draw(ht_hor)
draw(lgd, x = unit(w*0.45, "cm"), y = unit(0.8, "cm"), just = c("left", "bottom"))
dev.off()
```


# SPI2 analysis:
Here I perform a SPI2 analysis as described in the manuscript.

First, I read in the dataframes:
```{r}
topDEgenes <- c(rownames(pttt[pttt$p_value_FDR<0.001 & abs(pttt$logFC)>1,])[1:10],
                rownames(ptscrt[ptscrt$p_value_FDR<0.001 & 
                                           abs(ptscrt$logFC)>1,])[1:10],
                rownames(ttt[ttt$p_value_FDR<0.001 & 
                                           abs(ttt$logFC)>1,])[1:10], 
                rownames(pttr[pttr$p_value_FDR<0.001 & abs(pttr$logFC)>1,])[1:10],
                rownames(ptscrr[ptscrr$p_value_FDR<0.001 & 
                                           abs(ptscrr$logFC)>1,])[1:10],
                rownames(ttr[ttr$p_value_FDR<0.001 & 
                                           abs(ttr$logFC)>1,])[1:10],
                rownames(pttk[pttk$p_value_FDR<0.001 & abs(pttk$logFC)>1,])[1:10],
                rownames(ptscrk[ptscrk$p_value_FDR<0.001 & 
                                           abs(ptscrk$logFC)>1,])[1:10],
                rownames(ttk[ttk$p_value_FDR<0.001 & 
                                           abs(ttk$logFC)>1,])[1:10],
                "SL1344_1133", "SL1344_1134")
topDEgenes <- unique(topDEgenes[!is.na(topDEgenes)])
write_delim(data.frame(topDEgenes), "../data/PhoPQ_analysis_salcom/upgenes.txt"," ")

df_spi2 <- read.delim("../data/PhoPQ_analysis_salcom/salcom_query_degenes.txt", header = T, sep="\t")
df_phopq <- read.delim("../data/PhoPQ_analysis_salcom/salcom_query.phopq.txt", header = T, sep="\t")

df_salcom_spi2 <- data.frame(Wildtype = df_spi2$WT.MEP, SPI2 = df_spi2$WT.InSPI2, 
                        PhoPQ_ko = df_spi2$X.Delta.phoP.Q.InSPI2, row.names = df_spi2$SL1344.Locus.ID)

df_salcom_phopq <- data.frame(Wildtype = df_phopq$WT.MEP, SPI2 = df_phopq$WT.InSPI2, 
                        PhoPQ_ko = df_phopq$X.Delta.phoP.Q.InSPI2, row.names = df_phopq$SL1344.Locus.ID)

tpm_salcom_spi2 <- data.frame(sapply(df_salcom_spi2, function(x) as.integer(gsub(",","", x))), 
                              row.names = rownames(df_salcom_spi2))
tpm_salcom_phopq <- data.frame(sapply(df_salcom_phopq, function(x) as.integer(gsub(",","", x))), 
                              row.names = rownames(df_salcom_phopq))

tpm_salcom_spi2 <- tpm_salcom_spi2[order(tpm_salcom_spi2$SPI2),]
tpm_salcom_phopq <- tpm_salcom_phopq[order(tpm_salcom_phopq$SPI2),]
tpm_salcom_spi2 <- tpm_salcom_spi2[!(rownames(tpm_salcom_spi2)=="SL1344_1325"),]

tpm_salcom_spi2 <- tpm_salcom_spi2[!(rownames(tpm_salcom_spi2) %in% rownames(tpm_salcom_phopq)),]

tpm_salcom_spi2$condition <- "DE genes"
tpm_salcom_phopq$condition <- "PhoPQ"



rownames(tpm_salcom_phopq) <- gsub(".*MgrR", "MgrR", rownames(tpm_salcom_phopq))
tpm_salcom <- data.frame(rbind(tpm_salcom_phopq, tpm_salcom_spi2), row.names = c(rownames(tpm_salcom_phopq),
                                                                                 rownames(tpm_salcom_spi2)))

str(tpm_salcom)
```
Get all genes which are included in both datasets:
```{r}
all_spi2_genes <- rownames(tpm_salcom)[rownames(tpm_salcom) %in% rownames(y)]
str(all_spi2_genes)
tpm_salcom <- tpm_salcom[all_spi2_genes,]

prefname <- ifelse(all_spi2_genes %in% pnames$V2 ,pnames[all_spi2_genes,]$V1, "" )
prefname <- ifelse(isUnique(prefname), prefname, "")
prefname_spi2 <- ifelse(prefname != "", prefname, all_spi2_genes)

rownames(tpm_salcom) <- prefname_spi2
```

Now we have to make heatmaps with log foldchanges for all samples:
```{r}
# make list of all results, get logchange table:
reslist <- list(KFF = res_KFF,RXR = res_RXR, TAT = res_TAT)
logchanges_spi2 <- data.frame(lapply(reslist, function(l) {
  logfc <- data.frame( sapply(names(l), function(r) {
    l[[r]]$table[all_spi2_genes,1]
  }), row.names = all_spi2_genes )
})) 

colnames(logchanges_spi2) <- c("KFF-acpP ", "KFF-acpP-scrambled ", "KFF-only ","RXR-acpP ",
                          "RXR-acpP-scrambled ", "RXR-only ",
                          "Tat-acpP ", "Tat-acpP-scrambled ", "Tat-only")


# get mean cpm values of all conditions:
countscpm <- cpm(y)[all_spi2_genes,]

spi2_cpm <- sapply(levels(test), function(t) {
  rowMeans(countscpm[,t == test])
})
spi2_cpm <- spi2_cpm[,c(2,3,1,5,6,4,8,9,7,10)]


pvalues_spi2 <- data.frame(lapply(reslist, function(l) {
  logfc <- data.frame( sapply(names(l), function(r) {
    l[[r]]$table[all_spi2_genes,]$p_value_FDR<0.001 & abs(l[[r]]$table[all_spi2_genes,]$logFC)>1
  }), row.names = all_spi2_genes )
})) 

pvalues_spi2 <-sapply(pvalues_spi2, function(x) ifelse(x , x <- "*",x<-"")) 

```




Plot Heatmap:
```{r}
h <- dim(logchanges_spi2)[1] # width of plot (nr pws)

col_fun = colorRamp2(c(-2, 0, 2), c("blue", "beige", "red"))
ht_vert <- Heatmap(logchanges_spi2, cluster_rows = F, cluster_columns = F,
               name = "SPI2 analysis", col = col_fun,
               show_heatmap_legend = F, 
               row_title_side = "left", row_title_rot = 0,
               border = TRUE, 
               cell_fun = function(j, i, x, y, width, height, fill) {
                 grid.text(sprintf("%.1s", pvalues_spi2[i, j]), x, y)
               }, 
               column_names_gp = gpar(fontsize = 11),
               row_names_gp = gpar(fontsize = 10),
               column_split = factor(ord, levels = lev), 
               row_split = factor(tpm_salcom$condition),
               row_gap = unit(0.2, "cm"),
               width = unit(9*0.8, "cm"), height = unit(h/2, "cm"),
               column_names_rot = 45)


col_col_fun = colorRamp2(c(0, 2,3), c("lightgrey", "yellow", "red"))
ht_colgan <- Heatmap(log10(tpm_salcom[,1:3]), cluster_rows = F, cluster_columns = F,
               name = "Colgan et al., 2016", col = col_col_fun,
               show_heatmap_legend = F, 
               row_title_side = "right", row_title_rot = 0,
               border = TRUE, 
               row_split = factor(tpm_salcom$condition),
               column_names_gp = gpar(fontsize = 11),
               row_names_gp = gpar(fontsize = 10),
               row_gap = unit(0.2, "cm"),
               width = unit(3*0.8, "cm"), height = unit(h/2, "cm"),
               column_names_rot = 45)

order = factor(c(rep("KFF", 3), rep("RXR", 3), rep("TAT", 3), "Control"))



ht_list =  ht_vert + ht_colgan 
ht_list

lgd = Legend(col_fun = col_fun, title = expression("Log"[2]*" FC"),
             at = c(-2, 0, 2), legend_width = unit(4, "cm"),
             title_gp  = gpar(fontsize(14)),
             labels = c("-2", "  0", "  2"), legend_height = unit(3, "cm"),
             title_position = "leftcenter-rot")
lgd_col = Legend(col_fun = col_col_fun, title = expression("TPM"),
             at = c(0, 1, 2, 3), labels = c(0,10,100,1000), legend_width = unit(3, "cm"),
             title_gp  = gpar(fontsize(14)),
             legend_height = unit(3, "cm"),
             title_position = "leftcenter-rot")


pdf("../analysis/PhoPQ_Salcom/hm_salcom_phopq_spi2.pdf", width = 10, height = 15)
draw(ht_list, ht_gap = unit(0.5, "cm"))
draw(lgd, x = unit(4, "cm"), y = unit(18, "cm"), just = c("left", "bottom"))
draw(lgd_col, x = unit(21, "cm"), y = unit(18, "cm"), just = c("left", "bottom"))
dev.off()
```


save csv of all raw counts for supplementary material:
```{r}
counts <- y$counts 
test <- c("Water1", "KFF_acpP1", "KFF_acpP_scrambled1", "KFF1", "RXR_acpP1", "RXR_acpP_scrambled1", 
          "RXR1", "TAT_acpP1", "TAT_acpP_scrambled1", "TAT1",
          "Water2", "KFF_acpP2", "KFF_acpP_scrambled2", "KFF2", "RXR_acpP2", "RXR_acpP_scrambled2", 
          "RXR2", "TAT_acpP2", "TAT_acpP_scrambled2", "TAT2",
          "Water3", "KFF_acpP3", "KFF_acpP_scrambled3", "KFF3", "RXR_acpP3", "RXR_acpP_scrambled3", 
          "RXR3", "TAT_acpP3", "TAT_acpP_scrambled3", "TAT3")
colnames(counts) <- test

prefname <- ifelse(rownames(counts) %in% pnames$V2 ,pnames[rownames(counts),]$V1, "" )
prefname <- ifelse(isUnique(prefname), prefname, "")
gene_name <- ifelse(prefname != "", prefname, rownames(counts))

locus_tag <- rownames(counts)

counts_raw <- cbind(locus_tag, gene_name, counts)

write.csv(counts_raw, "../analysis/analysis_complete/supp_tables/raw_counts.csv")

for (i in all_res) {
  for (l in names(i)){
    tab <- i[[l]]$table[order(i[[l]]$table$FDR),]
    prefname <- ifelse(rownames(tab) %in% pnames$V2 ,pnames[rownames(tab),]$V1, "" )
    prefname <- ifelse(isUnique(prefname), prefname, "")
    genename <- ifelse(prefname != "", prefname, rownames(tab))
    tabs <- cbind(genename,tab) 
    write.csv(tabs, paste("../analysis/analysis_complete/supp_tables/",l,"_vs_water",".csv", sep = ""))
  }
}  


for (l in names(list_kegg_fry)){
  tabs <- list_kegg_fry[[l]]
  write.csv(tabs, paste("../analysis/analysis_complete/supp_tables/",l,"_vs_water",".csv", sep = ""))
}


```


Packages used:
```{r}
sessionInfo()
```

